portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -14,19 +14,83 @@
 
 package com.google.sps.servlets;
 
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.SortDirection;
+import com.google.gson.Gson;
+import com.google.sps.data.Comment;
 import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
 import javax.servlet.annotation.WebServlet;
 import javax.servlet.http.HttpServlet;
 import javax.servlet.http.HttpServletRequest;
 import javax.servlet.http.HttpServletResponse;
 
-/** Servlet that returns some example content. TODO: modify this file to handle comments data */
+/** Servlet that handles comments data. */
 @WebServlet("/data")
 public class DataServlet extends HttpServlet {
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    response.setContentType("text/html;");
-    response.getWriter().println("<h1>Hello world!</h1>");
+    Query query = new Query("Comment").addSort("timestamp", SortDirection.DESCENDING);
+    DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+
+    int limit = Integer.parseInt(request.getParameter("comment-limit"));
+    List<Entity> results = datastore.prepare(query).asList(FetchOptions.Builder.withLimit(limit));
+
+    List<Comment> comments = new ArrayList<>();
+    for (Entity entity : results) {
+      long id = entity.getKey().getId();
+      String name = (String) entity.getProperty("name");
+      String email = (String) entity.getProperty("email");
+      long timestamp = (long) entity.getProperty("timestamp");
+      String commentInput = (String) entity.getProperty("commentInput");
+
+      Comment comment = new Comment(id, name, email, timestamp, commentInput);
+      comments.add(comment);
+    }
+
+    response.setContentType("application/json;");
+    response.getWriter().println(new Gson().toJson(comments));
+  }
+
+  @Override
+  public void doPost(HttpServletRequest request, HttpServletResponse response) throws IOException {
```

Comment:
Function comment?
I'm a bit confused about the logic here: If comment-limit is set, then redirect; otherwise add it to the data store and redirect. Is it correct?

Reply:
+1 to helen's comment.

btw does it redirect because you're using the data servlet for both `<form action="/data" method="POST">` viewing number of comments as well as adding the comments?

```
<form action="/data" method="POST">
        <p>Name:</p>
        <input type="text" name="name">
        <br/>
        <p>Email address:</p>
        <input type="email" name="email">
        <br/>

        <p>Comment:</p>
        <textarea name="comment-input">Comment on blog...</textarea>
        <br/><br/>

        <input type="submit">
        <input type="reset">
        <br/><br/>
      </form>

      <h3>Comments</h3>
      <form action="/data" method="POST">
        <label for="comment-limit">View number of comments:</label>
        <input type="number" id="comment-limit" name="comment-limit" min="0" max="20">
      </form>
```

Reply:
(including the previous comment about comment-limit) so I wanted to make sure that when someone chooses to view a # of comments, they don't create another comment in the datastore (with the default values) b/c the form method is POST. I thought about changing the method to GET but felt that it wasn't a better option. Is there a better way of doing this (not giving the form a method maybe?) since the logic for the form is basically all in the JavaScript file?

Reply:
I'm fine with keeping it like this for now. You can just add some comment to explain the logic and reason.